### PR TITLE
Add view with all hosts

### PR DESCRIPTION
**Pull request by ar4s**

### CODE REVIEW

File: src/ralph/assets/admin.py [hunk 1]
```diff
@@ -273,6 +274,85 @@ def repr(self, obj):
...
+        # remove link
+        self.list_display_links = (None, )
+
+    def _initialize_search_form(self, extra_context, fields_from_model=True):
+        return super()._initialize_search_form(
+            extra_context, fields_from_model=False
+        )
+
+    def get_queryset(self, request):
+        # TODO: limit to DataCenterAsset, VirtualServer, CloudHost
```
Comment (ar4s) [inline src/ralph/assets/admin.py]: dd

File: src/ralph/assets/admin.py [hunk 2]
```diff
@@ -273,6 +274,85 @@ def repr(self, obj):
         return '{}: {}'.format(obj.content_type, obj)
 
 
+class BaseObjectList(BaseObject):
```
Comment (mkurek) [inline src/ralph/assets/admin.py]: move it to models

Comment (mkurek) [inline src/ralph/assets/admin.py]: and limit it to Cluster, DCAsset, CloudHost and VirtualServer

File: src/ralph/assets/admin.py [hunk 3]
```diff
@@ -273,6 +274,85 @@ def repr(self, obj):
...
+        'DataCenterAsset',
+        'VirtualServer'
+    ]
+
+    def get_actions(self, request):
+        return None
+
+    # TODO: hostname sorting
+    def get_hostname(self, obj):
+        return '<strong><a href="{}">{}</a></strong>'.format(
```
Comment (mkurek) [inline src/ralph/assets/admin.py]: consider subclassing `ChangeList` and overwriting `url_for_result`

File: src/ralph/assets/admin.py [hunk 4]
```diff
@@ -273,6 +274,85 @@ def repr(self, obj):
...
+        )
+    get_hostname.short_description = 'Hostname'
+    get_hostname.allow_tags = True
+
+    def __init__(self, model, *args, **kwargs):
+        super().__init__(model, *args, **kwargs)
+        # fixed issue with proxy model
+        self.opts = BaseObject._meta
+        # remove link
+        self.list_display_links = (None, )
```
Comment (mkurek) [inline src/ralph/assets/admin.py]: assign it on class level; use `list_display_links = None`

File: src/ralph/assets/admin.py [hunk 5]
```diff
@@ -273,6 +274,85 @@ def repr(self, obj):
...
+
+    def _initialize_search_form(self, extra_context, fields_from_model=True):
+        return super()._initialize_search_form(
+            extra_context, fields_from_model=False
+        )
+
+    def get_queryset(self, request):
+        # TODO: limit to DataCenterAsset, VirtualServer, CloudHost
+        qs = BaseObject.polymorphic_objects.dc_hosts().polymorphic_select_related(
```
Comment (mkurek) [inline src/ralph/assets/admin.py]: couldn't you just use regular `select_related`? afair it should be propagated to all polymorphic models

File: src/ralph/assets/models/base.py [hunk 6]
```diff
@@ -83,3 +83,33 @@ def _str_with_type(self):
...
+    ]
+    related_models = [
+        'Cluster',
+        'CloudHost',
+        'DataCenterAsset',
+        'VirtualServer'
+    ]
+
+    def get_queryset(self):
+        return BaseObject.polymorphic_objects.dc_hosts().polymorphic_select_related(
```
Comment (mkurek) [inline src/ralph/assets/models/base.py]: please check if Cluster is included in dc hosts

Comment (mkurek) [inline src/ralph/assets/models/base.py]: and check my previous comment about polymorphic and select related

File: src/ralph/assets/admin.py [hunk 7]
```diff
@@ -273,6 +278,59 @@ def repr(self, obj):
...
 
 
+class BaseObjectListChangeList(ChangeList):
+    def url_for_result(self, result):
+        return result.get_absolute_url()
+
+
+@register(BaseObjectList)
+class BaseObjectList(BaseObjectAdmin):
+    search_fields = [
```
Comment (mkurek) [inline src/ralph/assets/admin.py]: add ethernet__ipaddress__address

File: src/ralph/assets/models/base.py [hunk 8]
```diff
@@ -83,3 +84,24 @@ def _str_with_type(self):
             'class'
         )
     )
+
+
+class BaseObjectListManager(models.Manager):
+    select_related_fields = [
```
Comment (mkurek) [inline src/ralph/assets/models/base.py]: it's used by admin - move select related there

File: src/ralph/assets/models/base.py [hunk 9]
```diff
@@ -83,3 +84,24 @@ def _str_with_type(self):
...
+        'service_env__service'
+    ]
+
+    def get_queryset(self):
+        return BaseObject.polymorphic_objects.dc_hosts().select_related(
+            *self.select_related_fields
+        )
+
+
+class BaseObjectList(BaseObject):
```
Comment (mkurek) [inline src/ralph/assets/models/base.py]: BaseObjectList -> DCHost

Comment (mkurek) [inline src/ralph/assets/models/base.py]: and move it to data_center app

File: src/ralph/admin/filters.py [hunk 10]
```diff
@@ -458,6 +460,51 @@ def queryset(self, request, queryset):
...
+            return queryset
+        fields = [
+            'asset__hostname',
+            'cloudhost__hostname',
+            'cluster__hostname',
+            'virtualserver__hostname'
+        ]
+        # TODO: simple if hostname would be in one model
+        queries = [
+            Q(**{'{}__startswith'.format(field): self.value()})
```
Comment (mkurek) [inline src/ralph/admin/filters.py]: maybe icontains?

File: src/ralph/data_center/admin.py [hunk 11]
```diff
@@ -383,3 +389,64 @@ class DiskShareAdmin(RalphAdmin):
 @register(DiskShareMount)
 class DiskShareMountAdmin(RalphAdmin):
     pass
+
+
+class BaseObjectListChangeList(ChangeList):
```
Comment (mkurek) [inline src/ralph/data_center/admin.py]: -> DCHostChangeList

File: src/ralph/data_center/models/hosts.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    select_related_fields = [
+        'configuration_path',
+        'service_env',
+        'service_env__environment',
+        'service_env__service',
+    ]
+
+    def get_queryset(self):
+        return BaseObject.polymorphic_objects.dc_hosts().select_related(
+            *self.select_related_fields
```
Comment (mkurek) [inline src/ralph/data_center/models/hosts.py]: no need for select related here

File: src/ralph/admin/templates/admin/filter.html [hunk 13]
```diff
@@ -3,6 +3,7 @@
   {% blocktrans with filter_title=title %}
     By {{ filter_title }}
   {% endblocktrans %}
+  {{ choices }}
```
Comment (szok) [inline src/ralph/admin/templates/admin/filter.html]: It is necessary?

**Comment by szok** [comment]
:ok:

**Comment by szok** [comment]
:ok:

File: src/ralph/admin/filters.py [hunk 14]
```diff
@@ -473,6 +475,54 @@ def queryset(self, request, queryset):
...
+            (
+                ContentType.objects.get_for_model(model).pk,
+                model._meta.verbose_name
+            )
+            for model in models
+        ]
+        super().__init__(*args, **kwargs)
+
+
+class DCHostnameFilter(SimpleListFilter):
```
Comment (mkurek) [inline src/ralph/admin/filters.py]: -> DCHostHostnameFilter

File: src/ralph/admin/mixins.py [hunk 15]
```diff
@@ -169,12 +169,13 @@ def get_list_display_links(self, request, list_display):
         else:
             return None
 
-    def _initialize_search_form(self, extra_context):
-        search_fields = []
-        for field_name in self.search_fields:
-            field = get_field_by_relation_path(self.model, field_name)
-            search_fields.append(field.verbose_name)
-        extra_context['search_fields'] = search_fields
+    def _initialize_search_form(self, extra_context, fields_from_model=True):
```
Comment (mkurek) [inline src/ralph/admin/mixins.py:172]: describe `fields_from_model` param

File: src/ralph/admin/filters.py [hunk 16]
```diff
@@ -473,6 +475,54 @@ def queryset(self, request, queryset):
         return queryset
 
 
+class DCHostTypeListFilter(ChoicesListFilter):
```
Comment (mkurek) [inline src/ralph/admin/filters.py]: move it to `ralph.data_center.admin`

**Comment by mkurek** [comment]
:ok:

**Pull request merged**